packages/pn-commons/src/utils/notification.utility.ts

```diff
@@ -51,14 +53,19 @@ function localizeStatus(
   };
 }
 
+function hasTimelineElementForCategory(category: TimelineCategory, statusHistory: Array<NotificationStatusHistory>): boolean {
+  return statusHistory.some(status => status.steps?.some(step => step.category === category));
+}
+
+
 /**
  * Returns the mapping between current notification status and its color, label and descriptive message.
  * @param  {NotificationStatus} status
  * @returns object
  */
 export function getNotificationStatusInfos(
   status: NotificationStatus,
-  recipient?: string
+  options?: { recipient?: string; completeStatusHistory?: Array<NotificationStatusHistory> }
```

Comment:
In general I should avoid logic inside this function. There is a mapper function that do all the transformations before sending data to the component.
I should set a custom generic property in the NotificationStatusHistory for example:

```
export interface NotificationStatusHistory {
  status: NotificationStatus;
  activeFrom: string;
  relatedTimelineElements: Array<string>;
  // only fe
  steps?: Array<INotificationDetailTimeline>;
  details?: BaseDetails | DeliveredDetails;
}

interface BaseDetails {
 recipient?: string
}

interface DeliveredDetails extends BaseDetails {
  deliveringMode?: {enumeration for delivering modes}
}
```

and I populate this genric property in the mapper. In this way we'll put all transformations in one method and we'll avoid extra loops over notification object

Reply:
Even when I will an alternative solution following (I hope!) the lines you suggest, I strongly disagree with your view, since it concentrates the complexity in a single function (i.e. the mapper function) and data structure (NotificationStatusHistory) that becomes more and more difficult to understand as we add more details into it. 
In this particular case, we are adding an attribute in the general NotificationStatusHistory object, that makes sense only for the DELIVERED instance.
On the other hand, in my opinion performance gains that change only the quantity of operations but not the corresponding order (linear / logarithmic / polynomial / exponential) are of no practical benefit in the vast majority of cases, while isolating the individual tasks or aspects in different functions simplify the code where the gain in readibility is greater than the loss of (temporal) performance.
Indeed, I would rather go the other way around, namely in the direction of deflating populateMacroSteps and define separated functions for the diffent tasks done there, i.e. duplicating the ACCEPTED-related timeline elements hidden in one state and with no legal facts in the other one, setting recipient to the NotificationStatus.VIEWED state, and changing the VIEWED status to VIEWED_AFTER_DEADLINE if the notification passed through EFFECTIVE_DATE before.

In general, given a situation in which I can either pre-compute a datum and store it inside a data structure, or compute it in the place in which it is used, I tend to choose the second alternative, because it leads to more cohesive, less coupled and less complicated code. I taught this idea for 20 years and I'm still convinced of it.

Reply:
This is a transitional phase, in the near future the mapper function will be mooved to a lambda function in back-end.
So to have all the transformations in one place, it will help us to do this transition